### PR TITLE
[types/mempool] Update roles so that they are ranked in mempool

### PR DESCRIPTION
**Pull request by tzakian**
Right now we classify all transactions as either "governance" or "non-governance" where governance transactions are prioritized over all non-governance transactions. However, within governance there is no prioritization but we should, in particular LibraRoot and TC transactions should be prioritized over all other transactions including other governance-related transactions. 

This PR introduces this ranking for governance transactions.

This should not break client behavior when interacting with the system.

### CODE REVIEW

File: types/src/transaction/mod.rs [hunk 1]
```diff
@@ -592,16 +592,52 @@ impl From<VMStatus> for TransactionStatus {
...
+    }
+}
+
 /// The result of running the transaction through the VM validator.
 #[derive(Clone, Debug, Eq, PartialEq)]
 pub struct VMValidatorResult {
     status: Option<DiscardedVMStatus>,
     score: u64,
-    is_governance_txn: bool,
+    governance_role: Option<GovernanceRole>,
```
Comment (sblackshear) [inline types/src/transaction/mod.rs:632]: Thoughts on: 
- Having the `GovernanceRole` enum include all the roles (or one catch-all `NonGovernance` role)
- Getting rid of the `Option` here? 

I think that would let us simplify the comparison of `priority` in the mempool code and make it easier to test that `None` < `Some(GovernanceRole)`.

File: language/libra-vm/src/libra_transaction_validator.rs [hunk 2]
```diff
@@ -153,19 +150,22 @@ impl VMValidator for LibraVMValidator {
...
             res.map(|s| s.status_code()),
             normalized_gas_price,
-            is_prioritized_txn,
+            is_governance_txn,
         )
     }
 }
 
-fn is_prioritized_txn(sender: AccountAddress, remote_cache: &StateViewCache) -> bool {
+fn is_governance_transaction(
```
Comment (sblackshear) [inline language/libra-vm/src/libra_transaction_validator.rs:158]: Nit: perhaps a different name like `get_role`?

File: mempool/src/core_mempool/index.rs [hunk 3]
```diff
@@ -84,7 +84,12 @@ impl PartialOrd for OrderedQueueKey {
 
 impl Ord for OrderedQueueKey {
     fn cmp(&self, other: &OrderedQueueKey) -> Ordering {
-        match self.is_governance_txn.cmp(&other.is_governance_txn) {
+        match self
+            .governance_role
+            .as_ref()
+            .map(|x| x.priority())
+            .cmp(&other.governance_role.as_ref().map(|x| x.priority()))
```
Comment (sblackshear) [inline mempool/src/core_mempool/index.rs:91]: Will this do the right thing when one of the `governance_role`'s is `None`?

File: types/src/transaction/mod.rs (hunk 1 above)
Comment (tzakian) [reply types/src/transaction/mod.rs:632]: I like this idea a lot more than the option. I'll update and add a `NonGovernanceRole` variant.

File: mempool/src/core_mempool/index.rs (hunk 3 above)
Comment (tzakian) [reply mempool/src/core_mempool/index.rs:91]: It should. But I think it's another reason to get rid of the option :)

File: types/src/transaction/mod.rs [hunk 4]
```diff
@@ -592,16 +592,54 @@ impl From<VMStatus> for TransactionStatus {
...
+            0 => LibraRoot,
+            1 => TreasuryCompliance,
+            2 => DesignatedDealer,
+            3 => Validator,
+            4 => ValidatorOperator,
+            _ => NonGovernanceRole,
+        }
+    }
+
+    pub fn priority(&self) -> u64 {
```
Comment (christinacelee) [inline types/src/transaction/mod.rs:618]: nit: maybe a comment that the larger the priority value, the more greatly it is prioritized?

File: mempool/src/core_mempool/index.rs [hunk 5]
```diff
@@ -53,7 +53,7 @@ impl PriorityIndex {
             expiration_time: txn.expiration_time,
             address: txn.get_sender(),
             sequence_number: txn.get_sequence_number(),
-            is_governance_txn: txn.is_governance_txn,
+            governance_role: txn.governance_role.clone(),
```
Comment (christinacelee) [inline mempool/src/core_mempool/index.rs:56]: nit: maybe just derive `Copy` for GovernanceRole to remove cloning?

File: types/src/transaction/mod.rs [hunk 6]
```diff
@@ -592,16 +592,54 @@ impl From<VMStatus> for TransactionStatus {
...
+    LibraRoot,
+    TreasuryCompliance,
+    Validator,
+    ValidatorOperator,
+    DesignatedDealer,
+    NonGovernanceRole,
+}
+
+impl GovernanceRole {
+    pub fn from_role_id(role_id: u64) -> Self {
```
Comment (christinacelee) [inline types/src/transaction/mod.rs:606]: will the numeric role_id mappings be documented/maintained somewhere?

Comment (tzakian) [reply types/src/transaction/mod.rs:606]: Good question! yes. These are documented in the Move modules published on-chain (specifically Roles.move).

**Pull request closed without merging**